core/templates/dev/head/forms/formBuilder.js

```diff
@@ -747,9 +728,313 @@ oppia.config(['$provide', function($provide) {
+            },
+            data: function() {
+              var that = this;
+              // Set attributes of component according to data values.
+              customizationArgSpecs.forEach(function(spec) {
+                that.element.getChild(0).setAttribute(
+                  spec.name + '-with-value',
+                  oppiaHtmlEscaper.objToEscapedJson(
+                    that.data[spec.name] || ''));
+              });
+            },
+            init: function() {
+              editor.fire('lockSnapshot', {
+                dontUpdate: true
+              });
+              var that = this;
+              var isMissingAttributes = false;
+              // On init, read values from component attributes and save them.
+              customizationArgSpecs.forEach(function(spec) {
+                var value = that.element.getChild(0).getAttribute(
+                  spec.name + '-with-value');
+                if (value) {
+                  that.setData(
+                    spec.name, oppiaHtmlEscaper.escapedJsonToObj(value));
+                } else {
+                  isMissingAttributes = true;
+                }
+              });
+
+              if (!isMissingAttributes) {
+                // Need to manually $compile so the directive renders.
+                $compile($(this.element.$).contents())($rootScope);
+              }
+              $timeout(function() {
+                editor.fire('unlockSnapshot');
+                editor.fire('saveSnapshot');
+              });
+            }
+          });
+        }
+      });
+    });
+  }
+]);
+
+oppia.directive('ckEditorRte', [
+  'rteHelperService',
+  function(rteHelperService) {
+    return {
+      restrict: 'E',
+      scope: {
+        uiConfig: '&'
+      },
+      template: '<div><div></div>' +
+                '<div contenteditable="true" class="ckeditor-rte">' +
+                '</div></div>',
+      require: '?ngModel',
+      link: function(scope, el, attr, ngModel) {
+        var _RICH_TEXT_COMPONENTS = rteHelperService.getRichTextComponents();
+        var names = [];
+        var icons = [];
+        _RICH_TEXT_COMPONENTS.forEach(function(componentDefn) {
+          if (!(scope.uiConfig() &&
+                scope.uiConfig().hide_complex_extensions &&
+                componentDefn.isComplex)) {
+            names.push(componentDefn.name);
+            icons.push(componentDefn.iconDataUrl);
+          }
+        });
+
+        // Rules to prevent CKEDITOR from filtering components and wrappers.
+        var componentFilterRules = names.map(function(name) {
+          return 'oppia-noninteractive-' + name;
+        }).join(' ') + '(*)[*];';
+        var wrapperFilterRules = ' div span[type];';
+
+        var pluginNames = names.map(function(name) {
+          return 'oppia' + name;
+        }).join(',');
+        var buttonNames = [];
+        names.forEach(function(name) {
+          buttonNames.push('Oppia' + name);
```

Comment:
Bit of an odd button name? Don't you want e.g. the human-readable description, or is this meant to be an internal property of sorts?